services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2019 The Alcor Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+        you may not use this file except in compliance with the License.
+        You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+        Unless required by applicable law or agreed to in writing, software
+        distributed under the License is distributed on an "AS IS" BASIS,
+        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+
+package com.futurewei.alcor.dataplane.utils;
+
+import com.futurewei.alcor.dataplane.service.GoalStateService;
+import com.futurewei.alcor.schema.*;
+import com.futurewei.alcor.schema.Port.PortState;
+import com.futurewei.alcor.web.entity.dataplane.InternalPortEntity;
+import com.futurewei.alcor.web.entity.dataplane.InternalSubnetEntity;
+import com.futurewei.alcor.web.entity.dataplane.NeighborInfo;
+import com.futurewei.alcor.web.entity.dataplane.NetworkConfiguration;
+import com.futurewei.alcor.web.entity.port.PortEntity;
+import com.futurewei.alcor.web.entity.vpc.VpcEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+import static com.futurewei.alcor.schema.Port.PortConfiguration.FixedIp;
+
+@Component
+public class GoalStateManager {
+  @Autowired private GoalStateService goalStateService;
+  private static final Logger LOG = LoggerFactory.getLogger(GoalStateManager.class);
+  private Map<String, Set<String>> portsInSameSubnetMap = new HashMap<>();
+  private Map<String, Set<NeighborInfo>> neighborInfoInSameSubenetMap = new HashMap<>();
+
+  public Map<String, Goalstate.GoalState> transformNorthToSouth(
```

Comment:
The transformNorthToSouth needs major refactor. The methods has about 300 lines of codes. A general guidance is that a method should be limited to one page long, about up to 30 lines of codes.

If too long, it usually means that we don't break down the logic into small pieces.

Reply:
would do, first step is dcddbff

Reply:
after discussion in daily sync meeting, we would leave this for now due to changing logic

Reply:
It is fine. Create a tracking issue https://github.com/futurewei-cloud/alcor/issues/287.

Reply:
ok, then i would resolve this conversation